src/mon/Monitor.cc

```diff
@@ -550,11 +550,21 @@ void Monitor::handle_conf_change(const ConfigProxy& conf,
   if (changed.count("mon_health_to_clog") ||
       changed.count("mon_health_to_clog_interval") ||
       changed.count("mon_health_to_clog_tick_interval")) {
-    health_to_clog_update_conf(changed);
+    if (lock.is_locked_by_me()) {
```

Comment:
i'd suggest against using `lock.is_locked_by_me()`.  we should only use this function in `assert()`.

probably we should make `Monitor::lock` a recursive lock. @jecluis @gregsfortytwo what do you think?

Reply:
Recursive locks are usually a bad idea and we don’t use them elsewhere to solve these problems. Why don’t we know if we’re locked or not when this is invoked? If some of the config options require different behavior let’s put them in a different path.

Reply:
we do know. but `ConfigProxy` does (should) not know. there are two (indirect) callers of this function. 

1. `Monitor::ms_dispatch()`. which guards `_ms_dispatch(m)` with lock. the latter calls `ConfigProxy::injectargs()` indirectly. that's why the test failed in Sage's test.
1. `CephContext::do_command()`. it calls `ConfigProxy::call_observers()`, which does not hold the lock when calling `obs->handle_conf_change(*this, keys)`. 

from `ConfigProxy` 's point of view, it just wants to apply the latest changes. it should not be aware of if its caller is hold a lock of the observer or not at this moment. so, as another more invasive approach, we could add an (optional) parameter to `ConfigProxy::injectargs()` to indicate if the caller is holding a lock or not. the parameter will be passed all the way down to `md_config_obs_t::handle_conf_change()`.

Reply:
What's preventing us from adding a new lock to protect this code, rather than reusing the mon lock? I'm not a big fan of using the mon lock for everything (unless we want to limit new threads being handled by ms_dispatch()/_ms_dispatch() while we are doing something, or prevent other things to run in parallel for some reason).

If what we're trying here is to prevent the monitor and CephContext to compete over the same piece of code, then adding a new lock to the monitor class and grabbing that would be preferable IMO, rather than the monitor's giant lock.

Reply:
> What's preventing us from adding a new lock to protect this code, rather than reusing the mon lock?

@jecluis it's not very straightforward to add another lock for addressing this issue. please see the backtrace in http://tracker.ceph.com/issues/39625. because `Monitor::timer` is using `Monitor::lock` for protecting the events hooked to the timer when accessing/manipulating them. the `SafeTimer` always verify if the caller is holding the lock when changing `SafeTimer::events`.

if we want to protect timer without messing up with the grand lock, probably we could add a dedicated lock for `Monitor::timer`, and guard all caller sites of `timer.add_event*()` and `timer.cancel_event()` with this new lock.

Reply:
what about just dropping mon lock in the injectargs path?

Reply:
yeah, we could drop the mon lock in `prefix == "injectargs"` branch before `g_conf().injectargs()` and always acquire the lock in `Monitor::handle_conf_change()`.

Reply:
Well certainly calling ConfigProxy under the lock in ms_dispatch is breaking the interface contract as used elsewhere, so that’s where we we should resolve it. Either we figure out how to not hold the lock there (I think we can just drop it as suggested?) or else we develop a new interface for ConfigProxy that conveys this info

Reply:
> yeah, we could drop the mon lock in `prefix == "injectargs"` branch before `g_conf().injectargs()` and always acquire the lock in `Monitor::handle_conf_change()`.

yeah, i prefer this